blockstore.go

```diff
@@ -3,121 +3,308 @@ package lmdbbs
+
 type Blockstore struct {
+	// oplock is a two-tier concurrent/exclusive lock to synchronize mmap
+	// growth operations. The concurrent tier is used by blockstore operations,
+	// and the exclusive lock is acquired by the mmap grow operation.
+	oplock sync.RWMutex
+
+	// cursors.
+	cursorlock sync.Mutex
+	cursors    []*cursor
+
+	// dedupGrow deduplicates concurrent calls to grow(); it is recycled every
+	// time that grow() actually runs.
+	dedupGrow *sync.Once
+
+	// env represents a database environment. An lmdb env can contain multiple
+	// databases, all residing in the same shared memory map, but the blockstore
+	// only uses a 1:1 mapping between envs and dbs.
 	env *lmdb.Env
-	db  lmdb.DBI
+	// db is an object representing an LMDB database inside an env.
+	db lmdb.DBI
+	// opts are the options for this blockstore.
+	opts *Options
 
-	closed int32
+	pagesize int64 // the memory page size reported by the OS.
+	closed   int32
 }
 
 var (
 	_ blockstore.Blockstore = (*Blockstore)(nil)
 	_ blockstore.Viewer     = (*Blockstore)(nil)
 )
 
-func Open(path string) (*Blockstore, error) {
+type Options struct {
+	// Path is the directory where the LMDB blockstore resides. If it doesn't
+	// exist, it will be created.
+	Path string
+
+	// ReadOnly, if true, opens this blockstore in read-only mode.
+	ReadOnly bool
+
+	// NoSync disables flushing system buffers to disk immediately when
+	// committing transactions.
+	NoSync bool
+
+	// +++ DB sizing fields. +++ //
+	// InitialMmapSize is the initial mmap size passed to LMDB when
+	// creating/opening the environment.
+	InitialMmapSize int64
+
+	// MmapGrowthStepFactor determines the next map size when a write fails. The
+	// current size is multiplied by the factor, and rounded up to the next
+	// value divisible by os.Getpagesize(), to obtain the new map size, which is
+	// applied with mdb_env_set_mapsize.
+	MmapGrowthStepFactor float64
+
+	// MmapGrowthStepMax is the maximum step size by which we'll grow the mmap.
+	MmapGrowthStepMax int64
+	// --- DB sizing fields. --- //
+
+	// MaxReaders is the maximum amount of concurrent reader slots that exist
+	// in the lock table.
+	MaxReaders int
+}
+
+func Open(opts *Options) (*Blockstore, error) {
+	path := opts.Path
+	switch st, err := os.Stat(path); {
+	case os.IsNotExist(err):
+		if err := os.MkdirAll(path, 0777); err != nil {
+			return nil, fmt.Errorf("failed to create lmdb data directory at %s: %w", path, err)
+		}
+	case err != nil:
+		return nil, fmt.Errorf("failed to check if lmdb data dir exists: %w", err)
+	case !st.IsDir():
+		return nil, fmt.Errorf("lmdb path is not a directory %s", path)
+	}
+
+	pagesize := os.Getpagesize()
+	if pagesize < 0 {
+		pagesize = 4096 // being defensive and setting a safe value (4KiB).
+	}
+
+	// Validate mmap sizing parameters.
+	//
+	// InitialMmapSize cannot be negative nor zero, and must be rounded up to a multiple of the OS page size.
+	if v := opts.InitialMmapSize; v <= 0 {
```

Comment:
let's not require the user to set it if he wants just the defaults -- 0 should mean default.

Reply:
0 will already fall back to the default, but I can omit the warning on the log.

Reply:
good enough.